### ISSUE TITLE
Containerized running nats streaming cluster

### ISSUE THREAD

**Issue by sah4ez** [nats-io/nats-streaming-server#575]
Hi!
I try run nats-streaming-server in cluster in kubernetes. With storage type `sql`.
First I had this error https://github.com/nats-io/nats-streaming-server/issues/547
After I added `-cluster_log_path` in args for command, but nodes also didn't starting.

In `-cluster_log_path` I created `raft.log` file, but also got an error.
### version nats-streaming-server
```
0.9.2
```
### args of command run
```bash
 /opt/nats-streaming-server \
          -cid=MY_CID \
          -m=8222 \
          --clustered \
          --cluster_log_path=/var/log \
          -st=sql \
          --sql_driver=postgres \
          --sql_source=user=$(USER) password=$(PASSWORD) host=dev-postgres port=5432 dbname=$(DATABASE) sslmode=disable \
          --nats_server=nats://gnatsd:4222 \
          --cluster=nats://0.0.0.0:6222 \
          --routes=nats-route://nats-streaming:6222 \
          --max_msgs 0 \
          --max_bytes 0
```

### Error
```
[1] 2018/06/11 10:24:11.528873 [INF] STREAM: Starting nats-streaming-server[platoon] version 0.9.2
[1] 2018/06/11 10:24:11.528958 [INF] STREAM: ServerID: uactlWcb2bEYwOAeqDkxpK
[1] 2018/06/11 10:24:11.528966 [INF] STREAM: Go version: go1.9.5
[1] 2018/06/11 10:24:11.760455 [INF] STREAM: Recovering the state...
[1] 2018/06/11 10:24:11.772578 [INF] STREAM: Recovered 0 channel(s)
[1] 2018/06/11 10:24:11.772610 [INF] STREAM: Cluster Node ID : AsNO1G4hf0FQj4xFisO18d
[1] 2018/06/11 10:24:11.772613 [INF] STREAM: Cluster Log Path: /var/log
[1] 2018/06/11 10:24:11.776152 [INF] STREAM: Shutting down.
[1] 2018/06/11 10:24:11.970870 [FTL] STREAM: Failed to start: streaming state was recovered but cluster log path "/var/log" is empty 
```

**Comment by kozlovic** [comment]
As documented [here](https://github.com/nats-io/nats-streaming-server#clustering-and-containers), this happens if you start the server with clustered options for the first time with an existing Streaming store, which you can't do.

**Comment by sah4ez** [comment]
Ok. 
For first running nats-streaming-server with `-st=sql` I should use `--cluster_bootstrap`?
I try [this](https://github.com/nats-io/nats-streaming-server#clustering-auto-configuration) example with `-st=file`, and cluster was created successful. 

But with this command also got error:
```
/opt/nats-streaming-server \
          -cid=MY_CID \
          -m=8222 \
          --clustered \
          --cluster_bootstrap \
          --cluster_log_path=/var/log \
          -st=sql \
          --sql_driver=postgres \
          --sql_source=user=$(USER) password=$(PASSWORD) host=dev-postgres port=5432 dbname=$(DATABASE) sslmode=disable \
          --nats_server=nats://gnatsd:4222 \
          --cluster=nats://0.0.0.0:6222 \
          --routes=nats-route://nats-streaming:6222 \
          --max_msgs 0 \
          --max_bytes 0
```

What command need use for first running and initialisation cluster with sql-storage?

**Comment by kozlovic** [comment]
Regardless of FILE or SQL store, you need to either bootstrap the cluster, or define the explicit list of cluster peers. For instance, say that you have 3 nodes named `a`, `b` and `c`, for node `a` you would have these params `--cluster_node_id a --cluster_peers b,c` (note that you can specify the current node in the list of peers, see #522 and #523).

Also, when starting in `--clustered` mode, if the server detects no RAFT logs in the `--cluster_log_path` directory but find streaming stores (either files with `--dir` or SQL data), then it will fail because that means that there is pre-existing streaming data due to previous run without clustering mode, or mistakenly store the RAFT logs into non persistent storage.

In other words, you would need to start with an empty database and empty `cluster_log_path` the very first time.

**Comment by sah4ez** [comment]
> empty database 

It's mean for new database need execute [this script](https://github.com/nats-io/nats-streaming-server/blob/master/postgres.db.sql)?

**Comment by kozlovic** [comment]
Yes, but really the same way you first created your database when you ran without clustering.

In simple terms, it is not possible to upgrade from non clustering mode to clustering mode. So if you run the server in clustering mode for the first time, you can't do so if you are pointing to an existing datastore (file or sql).

**Comment by sah4ez** [comment]
When I run cluster in containers, `-cluster_log_path` directory should be shared between containers? And this directory not read only for nodes?

**Comment by ripienaar** [comment]
@sah4ez no that is not shared

**Comment by sah4ez** [comment]
@ripienaar but I ran this script and cluster successfully started::
```
#!/bin/bash

USER=postgres
PASSWORD=""
HOST=192.168.1.214
DATABASE=postgres

docker run -itd --name gnatsd -p 0.0.0.0:4222:4222 docker.scnetservices.ru/msg-gnatsd:0.1.4

docker run -itd --name nats-0 -p 0.0.0.0:6222:6222 \
  -v logs:/var/log \
  docker.scnetservices.ru/msg-nats-streaming:0.4.10 \
  /opt/nats-streaming-server -store=sql \
  -sql_driver=postgres \
  -cluster_bootstrap \
  -sql_source="user=postgres password='' host=192.168.1.214 port=5432 dbname=postgres sslmode=disable" \
  -clustered \
  -nats_server=nats://192.168.1.214:4222 \
  -cluster=nats://0.0.0.0:6222 \
  -routes=nats://192.168.1.214:6223,nats://192.168.1.214:6224 \
  -cluster_log_path=/var/log/

docker run -itd --name nats-1 -p 0.0.0.0:6223:6222 \
  -v logs:/var/log \
  docker.scnetservices.ru/msg-nats-streaming:0.4.10 \
  /opt/nats-streaming-server -store=sql \
  -sql_driver=postgres \
  -sql_source="user=postgres password='' host=192.168.1.214 port=5432 dbname=postgres sslmode=disable" \
  -clustered \
  -nats_server=nats://192.168.1.214:4222 \
  -cluster=nats://0.0.0.0:6223 \
  -routes=nats://192.168.1.214:6222,nats://192.168.1.214:6224 \
  -cluster_log_path=/var/log/

docker run -itd --name nats-2 -p 0.0.0.0:6224:6222  \
  -v logs:/var/log \
  docker.scnetservices.ru/msg-nats-streaming:0.4.10 \
  /opt/nats-streaming-server -store sql \
  -sql_driver=postgres \
  -sql_source="user=postgres password='' host=192.168.1.214 port=5432 dbname=postgres sslmode=disable" \
  -clustered \
  -nats_server=nats://192.168.1.214:4222 \
  -cluster=nats://0.0.0.0:6224 \
  -routes=nats://192.168.1.214:6222,nats://192.168.1.214:6223 \
  -cluster_log_path=/var/log/
```

If I removed `-v logs:/var/log`, then `node-1` and `node-2` didn't start, with error `log is empty`.

**Comment by kozlovic** [comment]
You need to specify a persistent place where the Streaming server will store the raft specific logs. If all nodes point to the same directory, that is fine, since in there, the server will create a directory with the name of the cluster and then a sub-directory per node id (either the explicit node id that you provide or a generated one). But a given node has its own directory, that's what @ripienaar meant in that the data stores do not need to be shared in clustering since data is replicated in each node (as opposed to the Fault Tolerance mode with active-standby and shared state).

**Comment by sah4ez** [comment]
@kozlovic If we say about deploying nats-streaming into Kubernetes cluster, that means for each pod need add `persistent volume` and `persistent volume claim`?

**Comment by kozlovic** [comment]
Yes, you need persistent volumes, because as explained [here](https://github.com/nats-io/nats-streaming-server#supported-stores) last sentence: "This means that even if you select an SQL Store, there will still be a need for storing data on the file system".

By the way, it looks like you configure routing between each NATS Streaming Server, so you don't need to start a standalone gnatsd (first docker command I see above), unless you use it for different purposes.

**Comment by juandiegopalomino** [comment]
@kozlovic I have a question about this setup and its monitoring. Just now, I (with minor tweaks) replicated sah4ez's script as follows

```
docker run -itd --name gnatsd -p 0.0.0.0:4222:4222 nats:latest

docker run -itd --name nats-0 -p 0.0.0.0:6222:6222 -p 0.0.0.0:8222:8222  \
  -v logs:/var/log \
  nats-streaming:latest \
  -store=sql -m=8222 \
  -sql_driver=postgres \
  -cluster_bootstrap \
  -sql_source="user=postgres password=secret host=172.17.0.1 port=30215 dbname=nats_streaming sslmode=disable" \
  -clustered \
  -nats_server=nats://172.17.0.1:4222 \
  -cluster=nats://0.0.0.0:6222 \
  -routes=nats://172.17.0.1:6223,nats://172.17.0.1:6224 \
  -cluster_log_path=/var/log/

docker run -itd --name nats-1 -p 0.0.0.0:6223:6222 -p 0.0.0.0:8223:8222 \
  -v logs:/var/log \
  nats-streaming:latest \
  -store=sql -m=8222 \
  -sql_driver=postgres \
  -sql_source="user=postgres password=secret host=172.17.0.1 port=30215 dbname=nats_streaming sslmode=disable" \
  -clustered \
  -nats_server=nats://172.17.0.1:4222 \
  -cluster=nats://0.0.0.0:6223 \
  -routes=nats://172.17.0.1:6222,nats://172.17.0.1:6224 \
  -cluster_log_path=/var/log/

docker run -itd --name nats-2 -p 0.0.0.0:6224:6222 -p 0.0.0.0:8224:8222  \
  -v logs:/var/log \
  nats-streaming:latest \
  -store sql -m=8222 \
  -sql_driver=postgres \
  -sql_source="user=postgres password=secret host=172.17.0.1 port=30215 dbname=nats_streaming sslmode=disable" \
  -clustered \
  -nats_server=nats://172.17.0.1:4222 \
  -cluster=nats://0.0.0.0:6224 \
  -routes=nats://172.17.0.1:6222,nats://172.17.0.1:6223 \
  -cluster_log_path=/var/log/
```
The main differences are some host & db access refactoring, updating the docker images to the canonical latest (as of a few minutes ago) as well as starting the monitoring endpoints. All the containers came up nicely, and seem to be listening as expected. I then submitted a handful of message over a couple of different channels with no problem.
The strangeness starts when I look at the monitoring endpoints: the channel/subscription endpoint of the leader, http://localhost:8222/streaming/channelsz?subs=1 is totally up to date and reporting the expected channel, messages, and subscriptions. The same endpoint of the slaves however, http://localhost:8223/streaming/channelsz?subs=1 & http://localhost:8224/streaming/channelsz?subs=1, report no messages in the system.
Is there something wrong with the setup, or anyway to ensure that the slaves' monitoring is up to date as well?

**Comment by kozlovic** [comment]
@juandiegopalomino See comments on issue #668 that you opened. Specific notes to your comments here:
- You are starting a gnatsd and specify `-nats_server` in each of the streaming server command line params, which means that all NATS Streaming servers will connect to that server. Therefore the `-cluster` and `-routes` parameters are not used. I would recommend that you have a look at [this](https://github.com/nats-io/nats-streaming-server#relation-to-nats) to better understand the relation to NATS. Also, there is a section in Clustering explaining how to configure each node when embedding NATS and a note about connecting to an existing NATS Server (or cluster of servers).
- As described in other issue, you are using the same database for each node, which is not the correct way. All servers will write their own data to the same database causing corruption.
